<a href="https://colab.research.google.com/github/mariron42/FisicaComputacional-1/blob/main/Actividad%203/%20act_3_py.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Act 3**: pandas y el clima 
---
Por Marcel Herrera Rendón 

***Paso 1***
---


---






**Importar librerías y cargar base de datos**

In [ ]:
import pandas as pd
import numpy as np

# Tomar la dirección URL del archivo en formato 'raw'
#Hacer una variable url para luego no tenener que escribir las la url abajo
url = 'https://raw.githubusercontent.com/mariron42/Los-trabajos-de-marcel/main/act3/2033.txt'






#variable que da nombre a las colunas 
my_cols = ['Fecha', 'Precip', 'Evap', 'Tmax', 'Tmin'] #Nombres de las columnas

df_dist = pd.read_csv(url, names=my_cols, encoding='cp1251', sep='\s+', header=None, skiprows=19, skipfooter=1, engine='python')
# El separador de datos es de uno o más espacios, sep='\s+'
# El archivo tiene códigos ASCII extendido (Español), encoding='cp1251' Windows (Europa Occidental)
# Se brincará los primeros 19 renglones y el último, afirmamos que no nay encabezado
# Se procesan los datos utilizando Python, engine='python' 


***Paso 3***
---


---





**Mostrar los primeros y ultimos 10 renglones de la base de datos**

In [ ]:
#Encabezado del dataframe(primeros 10 renglones)
print(df_dist.head(10))
#Final del dataframe(Ultimos 10 renglones)
print(df_dist.tail(10))

        Fecha Precip Evap  Tmax  Tmin
0  01/02/1944      0  1.2  23.7   4.8
1  02/02/1944      0  1.2  24.6   5.7
2  03/02/1944      0  2.8  21.8   7.4
3  04/02/1944      0  2.1  21.4  11.9
4  05/02/1944      0  2.2  21.2   8.4
5  06/02/1944      9  0.6  21.1   6.5
6  07/02/1944      0  1.2  21.7   5.7
7  08/02/1944      0    3  22.3   8.1
8  09/02/1944      0  2.3  21.9   7.9
9  10/02/1944      0  4.5  21.2   5.1
            Fecha Precip Evap  Tmax Tmin
23764  22/12/2012      0  1.6    19  2.5
23765  23/12/2012      0  1.9  19.8  4.5
23766  24/12/2012      0  1.4  23.4  5.5
23767  25/12/2012      0  4.5  17.6  6.5
23768  26/12/2012      0  3.1  19.8  4.5
23769  27/12/2012      0  3.3  17.6    8
23770  28/12/2012      0  2.6  17.8    5
23771  29/12/2012      0  2.1    18  3.5
23772  30/12/2012    0.5  1.6  16.4  3.5
23773  31/12/2012      0    2  16.4    4


**Tamaño del la base de datos, columnas y renglones (no kb o mb)**




In [ ]:
# Investigamos su tamaño con la función shape 
df_dist.shape
#(renglones, columnas)

(23774, 5)

**Información del dataframe**

In [ ]:
# Aquí dice el nombre de las columnas, numero de renglones en cada una y el uso de memoría

df_dist.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23774 entries, 0 to 23773
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   23774 non-null  object
 1   Precip  23774 non-null  object
 2   Evap    23774 non-null  object
 3   Tmax    23774 non-null  object
 4   Tmin    23774 non-null  object
dtypes: object(5)
memory usage: 928.8+ KB


**Respaldar el data frame**




In [ ]:
# Hacemos una copia de trabajo y preservamos el dataframe original
df_work0 = df_dist.copy()#Esta es la función para respaldar el data frame

df_work0.iloc[16400:16410] #con esto solo se muestra una parte de la base de datos donde hay nulos
#Nota: yo tuve que buscar esa parte

Fecha Precip  Evap  Tmax  Tmin
16400  01/09/1992      0  Nulo  37.2  20.6
16401  02/09/1992      0  Nulo  38.9  22.6
16402  03/09/1992      0  Nulo  39.1  21.8
16403  04/09/1992      0  Nulo  37.8  21.6
16404  05/09/1992      0  Nulo  38.3  21.9
16405  06/09/1992      0  Nulo  38.5  20.8
16406  07/09/1992      0  Nulo  38.6  21.5
16407  08/09/1992      0  Nulo  38.3  20.6
16408  09/09/1992      0  Nulo  40.9  17.8
16409  10/09/1992      0  Nulo  41.7  18.9

In [ ]:
# Reemplazamos los textos 'Nulo' por espacio en blanco ''
str_Nulo = "Nulo" 
#creamos la variable "str_Nulo"

df_work1 = df_work0.replace(to_replace=str_Nulo, value='', regex=True)

# Aquí es donde reamente se remplasa la variable


df_work1.iloc[16400:16410] #con esto solo se muestra una parte de la base de datos donde ya no hay nulos
#Nota: yo tuve que buscar esa parte
#Nota2: No olvidar usar work1 y no work0 me costó media hora darme cueta de eso.

Fecha Precip Evap  Tmax  Tmin
16400  01/09/1992      0       37.2  20.6
16401  02/09/1992      0       38.9  22.6
16402  03/09/1992      0       39.1  21.8
16403  04/09/1992      0       37.8  21.6
16404  05/09/1992      0       38.3  21.9
16405  06/09/1992      0       38.5  20.8
16406  07/09/1992      0       38.6  21.5
16407  08/09/1992      0       38.3  20.6
16408  09/09/1992      0       40.9  17.8
16409  10/09/1992      0       41.7  18.9

**Combertir valores de las variables**

In [ ]:
# Deseamos convertir los valores de las variables ambientales a núm. flotantes 
cols_list = ['Precip', 'Evap', 'Tmax', 'Tmin'] #Lista de columnas (se usará en el for)

#el for es una ciclo donde ejecutaremos nuestra función de conversión de cariables
for cols in cols_list:
  #Funcaión de conversión de variables
  df_work1[cols] = pd.to_numeric(df_work1[cols], errors='coerce') #to_numeric es la función 
  #(base de datos(lista de colunas), como llamar a los errores )
  #ese cols va a irse recorriendo en cada columna

df_work1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23774 entries, 0 to 23773
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Fecha   23774 non-null  object 
 1   Precip  23736 non-null  float64
 2   Evap    20731 non-null  float64
 3   Tmax    23740 non-null  float64
 4   Tmin    23737 non-null  float64
dtypes: float64(4), object(1)
memory usage: 928.8+ KB


**Sumar nulos**

In [ ]:
#Sabemos que hay valores que faltan (nosotros los creamos cuando remplazamos los nulos por "")
#Los sumamos
df_work1.isnull().sum()
#nomnredelabasededatos.sinvalores().sumar()

Fecha        0
Precip      38
Evap      3043
Tmax        34
Tmin        37
dtype: int64

In [ ]:
#Obcional del ejecutar
#Esto solo es para mostrar las variables como flotantes 
print(df_work1.head())
print(df_work1.head())

        Fecha  Precip  Evap  Tmax  Tmin
0  01/02/1944     0.0   1.2  23.7   4.8
1  02/02/1944     0.0   1.2  24.6   5.7
2  03/02/1944     0.0   2.8  21.8   7.4
3  04/02/1944     0.0   2.1  21.4  11.9
4  05/02/1944     0.0   2.2  21.2   8.4
        Fecha  Precip  Evap  Tmax  Tmin
0  01/02/1944     0.0   1.2  23.7   4.8
1  02/02/1944     0.0   1.2  24.6   5.7
2  03/02/1944     0.0   2.8  21.8   7.4
3  04/02/1944     0.0   2.1  21.4  11.9
4  05/02/1944     0.0   2.2  21.2   8.4


In [ ]:
# Un análisis estadístico breve de las variables numéricas redondeado a 3 cifras significativas
df_work1.describe().round(3)

Precip       Evap       Tmax       Tmin
count  23736.000  20731.000  23740.000  23737.000
mean       0.194      6.040     31.361     14.885
std        1.842      3.452      8.820      7.845
min        0.000      0.000      6.000     -8.000
25%        0.000      3.000     23.700      8.500
50%        0.000      5.700     31.400     14.100
75%        0.000      8.700     39.500     21.500
max       73.800     17.900     52.000     33.400

**Analisis**


1.   Tuvo una presipitación maxima de 73.800 milimetros mientras que si se sumara toda la presipitación registrada habeían 23.7 metros de agua.

2.   La temperatura minima fue 31 grados centigrados mientras que la minima fue de 14 C lo que se me hace muy raro porque buscando en internet he visto que ha llegado a los 5 grados.










**Trabajaremos la variable Fecha.**

Una variable tipo fecha, contiene mucha información condensada (Año, Mes, Día, Hora, Minuto, Segundo y fracciones de Segundo). Buscamos acceder a toda esa información para utilizarla.

Python tiene algunas herramientas para trabajar las fechas. Veamos algunas de ellas.

**Tercer respaldo**

In [ ]:
# Creamos una tercer data frame, el "df_work2".
df_work2 = df_work1.copy()
#Leer los primeros renglones de la columna fecha
df_work2['Fecha'].head()

0    01/02/1944
1    02/02/1944
2    03/02/1944
3    04/02/1944
4    05/02/1944
Name: Fecha, dtype: object

**Combertir la fecha a un formato que entieda python**

In [ ]:
# Pandas tiene la función pd.to_datetime que convierte una cadena en tipo datetime64[ns]
# Intentamos a ver si Python reconoce el formato en nuestro caso. 
# Agregamos la opción de que el primer dato es el día (dayfirst=True)
df_work2['Fecha']=pd.to_datetime(df_work2['Fecha'], dayfirst=True).copy()
print(df_work1.head())

        Fecha  Precip  Evap  Tmax  Tmin
0  01/02/1944     0.0   1.2  23.7   4.8
1  02/02/1944     0.0   1.2  24.6   5.7
2  03/02/1944     0.0   2.8  21.8   7.4
3  04/02/1944     0.0   2.1  21.4  11.9
4  05/02/1944     0.0   2.2  21.2   8.4


In [ ]:
#(Opcional para el funcionamiento)
# Verificamos que cada variable (columna), sea del tipo que esperamos.
# Observamos que la variable Fecha es de tipo 'datetime64[ns]'  
df_work2.dtypes

Fecha     datetime64[ns]
Precip           float64
Evap             float64
Tmax             float64
Tmin             float64
dtype: object

Agregar columnas

In [ ]:
# Buscamos ahora añadir dos columnas adicionales: df['Año'], df['Mes']
# Usaremos la función .dt.year y .dt.month para extraer el Año y Mes de la Fecha
df_work2['Año'] = df_work2['Fecha'].dt.year  #agregar columna año
df_work2['Mes'] = df_work2['Fecha'].dt.month #agregar columna mes

# **Ya tenemos todo hecho ahora solo probamos**

In [ ]:
# Verificamos que se hayan creado de forma correcta
# Tendremos ahora dos columnas adicionales que podemos utilizar para filtrar información.
#Tenemos mes, año, etc.
print(df_work2.head())
print(df_work2.tail())


       Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
0 1944-02-01     0.0   1.2  23.7   4.8  1944    2
1 1944-02-02     0.0   1.2  24.6   5.7  1944    2
2 1944-02-03     0.0   2.8  21.8   7.4  1944    2
3 1944-02-04     0.0   2.1  21.4  11.9  1944    2
4 1944-02-05     0.0   2.2  21.2   8.4  1944    2
           Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
23769 2012-12-27     0.0   3.3  17.6   8.0  2012   12
23770 2012-12-28     0.0   2.6  17.8   5.0  2012   12
23771 2012-12-29     0.0   2.1  18.0   3.5  2012   12
23772 2012-12-30     0.5   1.6  16.4   3.5  2012   12
23773 2012-12-31     0.0   2.0  16.4   4.0  2012   12


Verificar tipo de variables

In [ ]:
# Verificamos que Año y Mes son variables enteras.
# Hasta aquí llegaremos esta semana. 
df_work2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23774 entries, 0 to 23773
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Fecha   23774 non-null  datetime64[ns]
 1   Precip  23736 non-null  float64       
 2   Evap    20731 non-null  float64       
 3   Tmax    23740 non-null  float64       
 4   Tmin    23737 non-null  float64       
 5   Año     23774 non-null  int64         
 6   Mes     23774 non-null  int64         
dtypes: datetime64[ns](1), float64(4), int64(2)
memory usage: 1.3 MB
